contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -1070,15 +1070,70 @@ validate_special_function(char *func_nsname, char *func_name, int nargs, Oid *in
 
 	/* If function is not a special function no additional handling required */
 	if (special_func == NULL)
+	{
+		/* report error for case when NULL casted to different datatypes and passed as 2nd or 3rd argument of SUBSTRING() function */
+		if (strlen(func_name) == 9 && strncmp(func_name, "substring", 9) == 0)
+		{
+			for (int i = 1; i < nargs; i++)
+			{
+				if (input_typeids[i] != UNKNOWNOID)
+				{
+					Node *arg = (Node *) lfirst(list_nth_cell(fargs, i));
+					/* Throw error when input is constant and NULL */
+					if (IsA(arg, Const) && ((Const *)arg)->constisnull)
+					{
+						Datum                       tsql_typename;
+						char*						typ_name;
+						LOCAL_FCINFO(fcinfo, 1);
+
+						if (common_utility_plugin_ptr == NULL)
+							ereport(ERROR,
+									(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("Failed to find common utility plugin.")));
+
+						/* if tsql_typename is NULL it implies that inputTypId corresponds to UDT */
+						InitFunctionCallInfoData(*fcinfo, NULL, 0, InvalidOid, NULL, NULL);
+						fcinfo->args[0].value = ObjectIdGetDatum(input_typeids[i]);
+						fcinfo->args[0].isnull = false;
+						tsql_typename = (*common_utility_plugin_ptr->translate_pg_type_to_tsql) (fcinfo);
```

Comment:
I would suggest to create direct API of translate_pg_type_to_tsql instead of using SQL variant of it, because invoking SQL function using fcinfo is costly operation.

Reply:
Done